EQA CNN Pre-trainer for EQA baseline implementation

mukulkhanna opened the pull request:
## Motivation and Context

This is to implement the Encoder-Decoder CNN feature extractor to be used in the EQA baseline implementation, as in #374. Feature extraction from scene images is the first part of each of the subsequent trainers (VQA, PACMAN) in the EQA implementation.

Implementation based on **EmbodiedQA, Das et al, CVPR 2018** ([paper](https://embodiedqa.org/paper.pdf), [code](https://github.com/facebookresearch/EmbodiedQA))

As suggested by @erikwijmans, the idea is to break down #374 into smaller, more manageable PRs. 

Relevant issue: #369; More information in [README](https://github.com/mukulkhanna/habitat-api/blob/eqa-cnn-pretrain/habitat_baselines/il/README.md).

## How Has This Been Tested

By running all trainers manually. Most baseline_trainer tests don't apply here. I would need some help in curating relevant test cases.


<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!--- What types of changes does your code introduce? Leave all the items that apply: -->
- New feature (non-breaking change which adds functionality)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Thank you!

erikwijmans commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,507 @@
+    def train(self) -> None:
+        raise NotImplementedError
+
+    def eval(self) -> None:
```
I believe this is shared with `BaseRLTrainer`.   If so, this and that should be moved to the BaseTrainer.

erikwijmans replied:
I believe this is shared with `BaseRLTrainer`.   If so, this and that should be moved to the BaseTrainer.

erikwijmans commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,507 @@
+                        checkpoint_index=prev_ckpt_ind,
+                    )
+
+    def _setup_eval_config(self, checkpoint_config: Config) -> Config:
```
Same comment as with `eval`

mukulkhanna replied:
I have moved `eval` and `setup_eval_config` to the `BaseTrainer` in the latest commit.

erikwijmans commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,507 @@
+
+        return q_string
+
+    def _put_vqa_text_on_image(
```
Move to a utils file.

erikwijmans commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,507 @@
+
+        question = questions[0]
+
+        ckpt_epoch = ckpt_path[ckpt_path.rfind("/") + 1 :]
```
```suggestion
        ckpt_epoch = os.path.basename(ckpt_path)
```

erikwijmans commented on habitat_baselines/il/models/models.py:
```diff
@@ -0,0 +1,245 @@
+        return out_seg, out_depth, out_ae
+
+
+class MultitaskCNN(nn.Module):
```
This and `MultitaskCNNOutput` share a lot of code.  Maybe this should just have a function that does the second half of the forward pass of `MultitaskCNNOutput`?

mukulkhanna replied:
Done in the latest commit. Could you please have a look? Thanks

erikwijmans commented on habitat_baselines/il/trainers/eqa_cnn_pretrain_trainer.py:
```diff
@@ -0,0 +1,264 @@
+from habitat_baselines.common.base_il_trainer import BaseILTrainer
+from habitat_baselines.common.baseline_registry import baseline_registry
+from habitat_baselines.common.tensorboard_utils import TensorboardWriter
+from habitat_baselines.il.data.eqa_cnn_pretrain_data import (
```
Seems like the dataset didn't get committed

mukulkhanna replied:
Ya thanks, it got ignored because all `data` folders were specified in .gitignore. I have added it now.

erikwijmans commented on habitat_baselines/il/trainers/eqa_cnn_pretrain_trainer.py:
```diff
@@ -0,0 +1,264 @@
+        ae_loss = torch.nn.SmoothL1Loss()
+        seg_loss = torch.nn.CrossEntropyLoss()
+
+        np.random.seed(2)
```
Use the seed in the config.

erikwijmans commented on habitat_baselines/il/trainers/eqa_cnn_pretrain_trainer.py:
```diff
@@ -0,0 +1,264 @@
+            else torch.device("cpu")
+        )
+
+        assert torch.cuda.is_available(), "Cuda-enabled GPU required"
```
```suggestion
```

erikwijmans commented on habitat_baselines/il/trainers/eqa_cnn_pretrain_trainer.py:
```diff
@@ -0,0 +1,264 @@
+        )
+
+        assert torch.cuda.is_available(), "Cuda-enabled GPU required"
+        torch.cuda.set_device(config.TORCH_GPU_ID)
```
Only call this if CUDA is available.

mukulkhanna commented:
Thanks @erikwijmans for the comments. I have made the following suggested changes - 

- Merged `MultitaskCNNOutput` into `MultitaskCNN`
- Moved eval functions (`eval`, `setup_eval_config`) to `BaseTrainer`
- Moved function `put_vqa_text_on_image` to util
- Cleaned CUDA check during trainer's init
- Moved random seed to config file

erikwijmans commented on habitat_baselines/il/trainers/eqa_cnn_pretrain_trainer.py:
```diff
@@ -0,0 +1,261 @@
+        seg_loss = torch.nn.CrossEntropyLoss()
+
+        np.random.seed(config.EVAL_SEG_COLORS_SEED)
+        self.colors = np.random.randint(255, size=(41, 3))
```
Let's use the colors here: https://github.com/facebookresearch/habitat-sim/blob/master/habitat_sim/utils/common.py#L183

mukulkhanna replied:
d069071

mukulkhanna commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,223 @@
+class EQACNNPretrainDataset(Dataset):
+    """Pytorch dataset for Embodied Q&A's feature-extractor"""
+
+    def __init__(self, env, config, mode="train"):
```
Thanks for this suggestion @erikwijmans! I have made the changes.

mukulkhanna commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,223 @@
+                lmdb_env = self.val_lmdb
+
+        else:
+            lmdb_env = lmdb.open(
```
@erikwijmans We're going to need the transaction for the `__len__` function [here](https://github.com/mukulkhanna/habitat-api/blob/eqa-1/habitat_baselines/il/data/eqa_cnn_pretrain_data.py#L199), for which `lmdb_env`possibly should be opened in `__init__` at least once. 

I could open the env once, get the stat entries from the lmdb dataset, save them and then declare it = None (and close it before I open it again in `__getitem__`)? Can you think of a better way?

erikwijmans replied:
Yep, open it once, grab the length, then close.  Opening and stating and lmdb database is really cheap, so not worries there.

mukulkhanna replied:
Alright, done in bbb7a3a, thanks!

mukulkhanna commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,223 @@
+                        ep for ep in self.episodes if ep.episode_id == ep_id
+                    )
+
+                    pos_queue = episode.shortest_paths[0]
```
Thanks @erikwijmans. I've made the changes - b6ec1ad

erikwijmans commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,227 @@
+            seg[seg == -1] = 0
+            seg = seg.astype("uint8")
+
+            if random.random() < 0.8:
```
Didn't notice this before, but this is absolutely cannot be how train/val is split -- it would mean you are pre-training the CNN on the scenes we consider as held-out for evaluating the agent.  Please use the train-val split by scene that already exists.

mukulkhanna replied:
Thanks for pointing this out @erikwijmans . My inital thought was that the feature extractor could be agnostic to the train/test slipt for other tasks because feature extraction is more of a generic task. I'll make the changes. Thanks

erikwijmans replied:
Yeah, I see where you are coming from, but we can no longer call the val scenes held-out of the encoder was trained on them.

mukulkhanna replied:
Got it, thanks!

erikwijmans commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,227 @@
+            for ep in self.episodes:
+                if ep.scene_id not in self.scene_ids:
+                    self.scene_ids.append(ep.scene_id)
+                    self.scene_episode_dict[ep.scene_id] = [ep.episode_id]
```
Just add the episode instead of the episode ID to the dictionary.  That will simplify a lot of the logic.

mukulkhanna replied:
Done in bbb7a3a, thanks!

erikwijmans commented on habitat_baselines/il/README.md:
```diff
@@ -0,0 +1,68 @@
+
+### Pre-trained model
+
+Pre-trained EQA-CNN-Pretrain model can be downloaded from [here](https://drive.google.com/file/d/1bZbv6LVsxRMStZ7CjbnCFZIo1mnbuRXG/view?usp=sharing). 
```
This will need to be re-trained after the train/val issue is fixed.

erikwijmans commented on habitat_baselines/il/data/eqa_cnn_pretrain_data.py:
```diff
@@ -0,0 +1,227 @@
+                txn.put((sample_key + "_depth").encode(), depth.tobytes())
+                txn.put((sample_key + "_seg").encode(), seg.tobytes())
+
+    def get_frames(self, frames_path, num=0):
```
Unused method I think

mukulkhanna commented:
Thanks @erikwijmans. I've made the changes and simplified the code quite a bit.

Also put the model to train on the new dataset split. Will update the pretrained model link once done.

mukulkhanna commented on configs/tasks/eqa_mp3d.yaml:
```diff
@@ -9,7 +9,7 @@ ENVIRONMENT:
 SIMULATOR:
   AGENT_0:
-    SENSORS: ['RGB_SENSOR']
+    SENSORS: ['RGB_SENSOR', 'SEMANTIC_SENSOR', 'DEPTH_SENSOR']
```
Thanks @mathfac. But we're going to need the ground truth semantic maps for preparing the dataset. We'll also need it for calculating the loss values (+ outputting ground truth seg maps for qualitatively analyzing predicted maps) during eval.

mathfac replied:
Then it should be included in eval config, but not in base EQA config. Otherwise it will be available for baselines during training.

mukulkhanna replied:
Marking this as resolved since we discussed that currently there's no cleaner way to have different sensor configs for train and eval.

mukulkhanna commented on habitat_baselines/common/utils.py:
```diff
@@ -228,6 +229,93 @@ def generate_video(
         )
 
 
+def tensor_to_depth_images(tensor: Union[torch.Tensor, List]) -> np.ndarray:
```
Thanks @mathfac. I'll create `habitat_baselines/utils/visualizations` for the same. 

Would you also want me to move other utils from `habitat_baselines/common` (like `env_utils`) also to `habitat_baselines/utils/` ?

mathfac replied:
that should be good here.

mukulkhanna replied:
I'll open a new PR for porting these utils to `habitat_baselines/utils/`?

mukulkhanna commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,379 @@
+        prediction: str,
+        ground_truth: str,
+    ) -> None:
+        r"""For saving image results.
```
Moved this to `utils/visualization/utils.py` as `save_vqa_image_results` and added more info in doc string.

mukulkhanna commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,379 @@
+            image,
+        )
+
+    def _save_vqa_results(
```
Thanks @mathfac. I'll add this to the VQA trainer when I push code for that. I've removed it from here.

mukulkhanna commented on habitat_baselines/common/base_il_trainer.py:
```diff
@@ -0,0 +1,379 @@
+                    split="val", type=type
+                )
+                if not os.path.isdir(dir_name):
+                    os.makedirs(dir_name)
```
Moved to derived trainer.

mukulkhanna commented:
Thank you for the suggestions and feedback @mathfac. 

I've now moved all EQA-dedicated functions away from the BaseILTrainer into either the derived trainer or baseline utils. I've also created the visualization utils for habitat baselines. 

I've added comments and marked many of the suggestions as resolved. Please let me know of any further changes required. Thanks!

mathfac commented:
Tests failed:
`test/test_examples.py` 
Also lint isn't happy, can you try use pre-commit  hook: `pip install pre-commit && pre-commit install.`
Thank you!

mathfac commented:
Thank you! Can you share command lines I can use to test the PR with MP3D data downloaded?

mathfac commented on habitat_baselines/common/utils.py:
```diff
@@ -228,6 +228,61 @@ def generate_video(
+    return images
+
+
+def get_q_string(question: List, q_vocab_dict: Dict) -> str:
```
Can we use VocabDict's function here:

https://github.com/facebookresearch/habitat-lab/blob/89cf9842259c219418c9ad185bc2d26e7d1e68c4/habitat/datasets/utils.py#L85

Maybe even add a method `token_idx_2_string` to VocabDict will be more re-usable. Then you can call `q_vocab_dict.token_idx_2_string(question) + "?"` instead  `get_q_string(question, q_vocab_dict)`

mukulkhanna replied:
Thanks @mathfac. 

I'm removing all VQA related utils from this PR. I'll address this in a subsequent PR for the VQA baseline implementation.

mathfac commented:
Everything else looks great!

mukulkhanna commented:
Thanks @mathfac!

I've fixed the linting, and am looking into the failed test cases.

> Thank you! Can you share command lines I can use to test the PR with MP3D data downloaded?

I have provided the train and eval commands (along with a pre-trained model and other info) in the README [here](https://github.com/mukulkhanna/habitat-api/blob/eqa-cnn-pretrain/habitat_baselines/il/README.md#train).

**Train**:
 ```
python -u habitat_baselines/run.py --exp-config habitat_baselines/config/eqa/il_eqa_cnn_pretrain.yaml --run-type train
```

**Eval**:
```
python -u habitat_baselines/run.py --exp-config habitat_baselines/config/eqa/il_eqa_cnn_pretrain.yaml --run-type eval
```

erikwijmans commented:
Looks like we need to add `lmdb` to the requirements

mukulkhanna commented:
Thanks @erikwijmans and @mathfac . I've made the suggested changes. Please let me know if there's anything else. Thanks!

mathfac reviewed (approved):
Great job @mukulkhanna! Thank you for persistence and finalizing this PR. Sure it will be speed up research projects with easy baselines comparison.
